nilearn/connectome/connectivity_matrices.py

```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
 
 
     """
     if discard_diagonal:
         # No scaling, we directly return the values
-        tril_mask = np.tril(np.ones(symmetric.shape[-2:]), k=-1).astype(np.bool)
+        tril_mask = np.tril(np.ones(symmetric.shape[-2:]), k=-1).astype(
+            np.bool)
         return symmetric[..., tril_mask]
     scaling = np.ones(symmetric.shape[-2:])
     np.fill_diagonal(scaling, sqrt(2.))
     tril_mask = np.tril(np.ones(symmetric.shape[-2:])).astype(np.bool)
     return symmetric[..., tril_mask] / scaling[tril_mask]
 
 
+def vec_to_sym_matrix(vec, diagonal=None):
+    """Return the symmetric matrix given its flattened lower triangular part.
+    Acts on the last dimension of the array if not 1-dimensional.
+    Diagonal can be encompassed in vec or given separately. In both cases, note
+    that diagonal elements are multiplied by sqrt(2).
+
+    .. versionadded:: 0.2.6.1
+
+    Parameters
+    ----------
+    vec : numpy.ndarray, shape (..., n_columns * (n_columns + 1) /2) or
+        (..., (n_columns - 1) * n_columns / 2) if diagonal is given seperately.
+        The input array.
+
+    diagonal : numpy.ndarray, shape (..., n_columns), optional
+        The diagonal array to be stacked to vec.
+
+    Returns
+    -------
+    sym : numpy.ndarray, shape (..., n_columns, n_columns).
+        The output symmetric matrix.
+
+    Note
+    ----
+    This function is meant to be the inverse of sym_matrix_to_vec. If you have
+    discarded the diagonal in sym_matrix_to_vec, you need to provide it
+    separately to reconstruct the symmetric matrix. For instance this can be
+    useful for correlation matrices for which we know the diagonal is 1.
+
+    See also
+    --------
+    nilearn.connectome.sym_matrix_to_vec
+    """
+    n = vec.shape[-1]
+    # Compute the number of the symmetric matrix columns
+    # solve n_columns * (n_columns + 1) / 2 = n subject to n_columns > 0
+    n_columns = (sqrt(8 * n + 1) - 1.) / 2
+    if diagonal is not None:
+        n_columns += 1
+
+    if n_columns > floor(n_columns):
+        raise ValueError(
+            "Vector of unsuitable shape {0} can not be transformed to "
+            "a symmetric matrix.".format(vec.shape))
+
+    n_columns = int(n_columns)
+    first_shape = vec.shape[:-1]
+    if diagonal is not None:
+        if diagonal.shape[:-1] != first_shape or\
+                diagonal.shape[-1] != n_columns:
+            raise ValueError("diagonal of shape {0} incompatible with vector "
+                             "of shape {1}".format(diagonal.shape, vec.shape))
+
+        # Include diagonal elements in vec
+        diagonal_mask = np.array(np.eye(n_columns), dtype=bool)
+        diagonal_mask = sym_matrix_to_vec(diagonal_mask).astype(np.bool)
+        vec_with_diagonal = np.zeros(first_shape + (n + n_columns,))
+        vec_with_diagonal[..., diagonal_mask] = diagonal
+        vec_with_diagonal[..., np.logical_not(diagonal_mask)] = vec
+        return vec_to_sym_matrix(vec_with_diagonal)
```

Comment:
Reallocating the whole data just to fill one diagonal? This is not necessary and can be done further.